docs/block-processing/types.md

```diff
@@ -0,0 +1,951 @@
+is defined differently:
+
+```haskell
+type VssCertificatesHash = Hash (HashMap StakeholderId VssCertificate)
+```
+
+The reason for that is that hashing is done after serialization, and at some
+point the serialization format for `VssCertificatesHash` was changed from a
+map to a set. Since we can't change the protocol easily at this point, for
+hashing we still use the map representation (where the map key corresponding
+to each certificate is a `StakeholderId` derived from its the certificate's
+`vcSigningKey`).
+
+#### Delegation payload+proof
+
+*keywords: `DlgPayload`*
+
+The delegation payload is a list of proxy keys. The invariant is that no two
+proxy keys have the same issuer (`pskIssuerPk`).
+
+```haskell
+newtype DlgPayload = UnsafeDlgPayload [ProxySKHeavy]
+```
+
+The proof used for this type is simply a `Hash DlgPayload`.
+
+#### Update system payload+proof
+
+*keywords: `UpdatePayload`, `UpdateProof`*
+
+The update system payload contains a list of votes and (optionally) an
+update proposal.
+
+```haskell
+data UpdatePayload = UpdatePayload
+    { upProposal :: Maybe UpdateProposal
+    , upVotes    :: [UpdateVote]
+    }
+```
+
+This type has no invariants.
+
+The proof used for this type is simply a `Hash UpdatePayload`.
+
+## Transaction-related types
+
+### Transaction
+
+*keywords: `Tx`, `TxId`, `TxIn`, `TxOut`, `TxAttributes`*
+
+A transaction (`Tx`) can be thought of as a command to destroy several
+*unspent outputs* and create new unspent outputs, possibly with different
+owners. In this fashion, money can be transferred from one person to another
+by destroying money belonging to the first person and creating money that
+would belong to the second person. (See [Transactions in Cardano SL][] for
+more details.)
+
+[Transactions in Cardano SL]: https://cardanodocs.com/cardano/transactions/
+
+Here is the definition of a transaction:
+
+```haskell
+data Tx = UnsafeTx
+    { _txInputs     :: NonEmpty TxIn         -- inputs
+    , _txOutputs    :: NonEmpty TxOut        -- outputs
+    , _txAttributes :: TxAttributes          -- attributes
+    }
+```
+
+The invariant is that all outputs must have value bigger than 0. Two
+additional invariants (“there's at least one input” and “there's at least
+one output”) are always satisfied thanks to using `NonEmpty`.
+
+Transactions are referred to by their hashes:
+
+```haskell
+type TxId = Hash Tx
+```
+
+----------------------------------------
+
+Inputs are represented like this: since currently unspent outputs can only
+originate from transactions, any unspent output can be referred to by
+specifying a transaction ID and index in the list of outputs:
+
+```haskell
+data TxIn
+    = TxInUtxo
+    { txInHash  :: TxId
+    , txInIndex :: Word32
+    }
+    | TxInUnknown Word8 ByteString
+```
+
+The `TxInUnknown` constructor is needed to be able to extend `TxIn` in the
+future without breaking backwards compatibility. TODO: link wherever I
+should link
```

Comment:
Perhaps it's better to describe it here and then to this place from another place.